test/test_fmm.py

```diff
@@ -51,36 +51,46 @@
-    (HelmholtzKernel(2), VolumeTaylorLocalExpansion, VolumeTaylorMultipoleExpansion,
-                         False),
-    (HelmholtzKernel(2), LinearPDEConformingVolumeTaylorLocalExpansion,
-                         LinearPDEConformingVolumeTaylorMultipoleExpansion, False),
-    (HelmholtzKernel(2), H2DLocalExpansion, H2DMultipoleExpansion, False),
-    (HelmholtzKernel(3), VolumeTaylorLocalExpansion, VolumeTaylorMultipoleExpansion,
-                         False),
-    (HelmholtzKernel(3), LinearPDEConformingVolumeTaylorLocalExpansion,
-                         LinearPDEConformingVolumeTaylorMultipoleExpansion, False),
-    (YukawaKernel(2), Y2DLocalExpansion, Y2DMultipoleExpansion, False),
+@pytest.mark.parametrize(
+    "knl, local_expn_class, mpole_expn_class, optimized_m2l, use_fft",
+    [
+        (LaplaceKernel(2), VolumeTaylorLocalExpansion,
+            VolumeTaylorMultipoleExpansion, False, False),
+        (LaplaceKernel(2), VolumeTaylorLocalExpansion,
+            VolumeTaylorMultipoleExpansion, True, False),
+        (LaplaceKernel(2), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, True, False),
+        (LaplaceKernel(2), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, True, True),
+        (LaplaceKernel(2), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, False, False),
+        (LaplaceKernel(3), VolumeTaylorLocalExpansion,
+            VolumeTaylorMultipoleExpansion, False, False),
+        (LaplaceKernel(3), VolumeTaylorLocalExpansion,
+            VolumeTaylorMultipoleExpansion, True, False),
+        (LaplaceKernel(3), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, True, False),
+        (LaplaceKernel(3), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, False, False),
+        (HelmholtzKernel(2), VolumeTaylorLocalExpansion,
+            VolumeTaylorMultipoleExpansion, False, False),
+        (HelmholtzKernel(2), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, False, False),
+        (HelmholtzKernel(2), H2DLocalExpansion,
+            H2DMultipoleExpansion, False, False),
+        (HelmholtzKernel(2), H2DLocalExpansion,
+            H2DMultipoleExpansion, True, False),
+        #(HelmholtzKernel(2), H2DLocalExpansion,
+        #    H2DMultipoleExpansion, True, True),
+        (HelmholtzKernel(3), VolumeTaylorLocalExpansion,
+            VolumeTaylorMultipoleExpansion, False, False),
+        (HelmholtzKernel(3), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, False, False),
+        (YukawaKernel(2), Y2DLocalExpansion,
+            Y2DMultipoleExpansion, False, False),
```

Comment:
*cough* loops *cough*?

Reply:
Fixed in 22e493063d70be8726dc72c41e855989b6c46789